File: Introduction_to_MPI/Solutions/c/ex3-ring/ring.c

/******************************************************************************
 *
 * Accumulate the sum of the ranks by sending a message around a
 * ring.
 *
 *  We use a synchronous non-blocking send mpi_issend() and a non-
 *  blocking recieve mpi_irecv(), along with two wait statements.
 *
 *****************************************************************************/

#include <mpi.h>
#include <stdio.h>

int main (int argc, char* argv[]){

  int rank, size;
  int my_prev, my_next;
  int sum;

  /*  Message information  */
  int sendbuf, recvbuf;
  int my_tag = 9999;
  MPI_Request reqr;
  MPI_Request reqs;
  MPI_Status status[1];

  /* loop counter */
  int n;

  MPI_Init(&argc,&argv);

  MPI_Comm_rank(MPI_COMM_WORLD, &rank);
  MPI_Comm_size(MPI_COMM_WORLD, &size);

  /* Work out neighbouring ranks */
  my_prev = rank - 1;
  
  if (my_prev < 0) 
  	my_prev = my_prev + size;
  
  my_next = (rank + 1) % size;

  /* Initialise the sum and the message for all ranks */
  sum = 0;
  sendbuf = rank;
        
  for (n = 0; n < size; n++) {
    MPI_Irecv(&recvbuf, 1, MPI_INT, my_prev, my_tag, MPI_COMM_WORLD, &reqr);
    MPI_Issend(&sendbuf, 1, MPI_INT, my_next, my_tag, MPI_COMM_WORLD, &reqs);

    /* When the receive has completed, we can use the contents of recvbuf */
    MPI_Wait(&reqr, status);
    sum = sum + recvbuf;

    /* When the send has completed, we can safely re-use the send buffer */
    MPI_Wait(&reqs, status);
    sendbuf = recvbuf;
  }

  /* Display the result on all ranks, along with the correct answer */

  printf("Rank %2d has sum of ranks %d; Answer = %d\n",rank, sum, (size-1)*size/2);

  MPI_Finalize();
	
  return 0;

} 

File: Introduction_to_MPI/Solutions/fortran/ex3-ring/ring.f90
program ring

  ! Accumulate the sum of the ranks by sending a message around a
  ! ring.

  ! We use a synchronous non-blocking send mpi_issend() and a non-
  ! blocking recieve mpi_irecv(), along with two wait statements.

  use mpi
  implicit none

  integer :: ifail
  integer :: rank, size
  integer :: my_prev, my_next
  integer :: sum

  ! Message information

  integer :: sendbuf, recvbuf
  integer :: req1, req2
  integer, parameter :: my_tag = 9999
  integer, dimension(MPI_STATUS_SIZE) :: status

  ! loop counter
  integer :: n

  call mpi_init(ifail)

  call mpi_comm_rank(MPI_COMM_WORLD, rank, ifail)
  call mpi_comm_size(MPI_COMM_WORLD, size, ifail)

  ! Work out neighbouring ranks

  my_prev = rank - 1
  if (my_prev < 0) my_prev = my_prev + size
  my_next = mod(rank + 1, size)

  ! Initialise the sum and the message for all ranks
  sum = 0
  sendbuf = rank

  do n = 1, size

     call mpi_irecv(recvbuf, 1, MPI_INTEGER, my_prev, my_tag, MPI_COMM_WORLD, &
                    req1, ifail)
     call mpi_issend(sendbuf, 1, MPI_INTEGER, my_next, my_tag, MPI_COMM_WORLD, &
                     req2, ifail)

     ! When the receive has completed, we can use the contents of recvbuf
     call mpi_wait(req1, status, ifail)
     sum = sum + recvbuf

     ! When the send has completed, we can safely re-use the send buffer
     call mpi_wait(req2, status, ifail)
     sendbuf = recvbuf

  end do

  ! Display the result on all ranks, along with the correct answer

  write (unit = *, fmt = '(a,i3,a,i4,a,i4)') &
       "Rank ", rank, " has sum of ranks ", sum, " Answer ", (size - 1)*size/2

  call mpi_finalize(ifail)


end program ring
